### PR TITLE
feat:Cluster description note

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description

Develop GET and PUT methods of Cluster Description.
Cluster Description Fields :
- ClusterName
- ClusterCreatedOn (Time Stamp)
- ClusterCreatedBy (user email id)
- ClusterNote Fields :
  - Description (Markup Text)
  - CreatedBy (user email id)
  - CreatedOn (Time Stamp)

Maintain audit history of cluster description note in database.

Ticket [AB 2434](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/OSS%20adoption/Stories/?workitem=2434)
[Tech Doc](https://docs.google.com/document/d/1xzYU5LUEf2cAsccKFQQeBtwu1UkR-VTFjFeFueiJ8Wk/edit?usp=sharing)

## How Has This Been Tested?

**GET:** /description?id={clusterId}

- [x] RBAC implementation for cluster get access.
- [x] return 404 for invalid clusterId.
- [x] Always Return clusterName, clusterCreatedOn, clusterCreatedBy if cluster exists.
- [x] Return NULL clusterNotes for none existing cluster notes.

**PUT:** /description/note

- [x] RBAC implementation for cluster update access.
- [x] return 404 for invalid clusterId
- [x] Return updated cluster notes on successful updation.
- [x] Audit gets created on CREATE and UPDATE methods implemented on cluster notes.

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/cluster/repository/ClusterNoteRepository.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	clusterNote := &ClusterNote{}
+	err := impl.dbConnection.
+	Model(clusterNote).
+	Where("cluster_id =?", id).
+	Limit(1).
+	Select()
+	return clusterNote, err
+}
+
+func (impl ClusterNoteRepositoryImpl) FindByClusterIds(id []int) ([]ClusterNote, error) {
```
Comment (vikramdevtron) [inline pkg/cluster/repository/ClusterNoteRepository.go:68]: use Ids in params

File: pkg/cluster/repository/ClusterNoteHistoryRepository.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+type ClusterNoteHistoryRepositoryImpl struct {
+	dbConnection *pg.DB
+	logger       *zap.SugaredLogger
+}
+
+func (impl ClusterNoteHistoryRepositoryImpl) SaveHistory(model *ClusterNoteHistory) error {
+	return impl.dbConnection.Insert(model)
+}
+
+func (impl ClusterNoteHistoryRepositoryImpl) FindHistoryByNoteId(id []int) ([]ClusterNoteHistory, error) {
```
Comment (vikramdevtron) [inline pkg/cluster/repository/ClusterNoteHistoryRepository.go:55]: user pointer * array

File: pkg/cluster/ClusterNoteService.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	}
+	return bean, err
+}
+
+func (impl *ClusterNoteServiceImpl) Delete(bean *ClusterNoteBean, userId int32) error {
+	model, err := impl.clusterNoteRepository.FindByClusterId(bean.ClusterId)
+	if err != nil {
+		return err
+	}
+	return impl.clusterNoteRepository.Delete(model)
```
Comment (vikramdevtron) [inline pkg/cluster/ClusterNoteService.go:207]: avoid hard delete

File: pkg/cluster/ClusterNoteService.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	}
+	bean.Id = model.Id
+	// audit the existing description to cluster audit history
+	clusterAudit := &repository.ClusterNoteHistory{
+		NoteId:      bean.Id,
+		Description: bean.Description,
+	}
+	clusterAudit.CreatedBy = userId
+	clusterAudit.UpdatedBy = userId
+	clusterAudit.CreatedOn = time.Now()
```
Comment (vikramdevtron) [inline pkg/cluster/ClusterNoteService.go:193]: on update why created by and created on ?

File: pkg/cluster/repository/ClusterNoteRepository.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	Where("cluster_id in(?)", pg.In(id)).
+	Select()
+	return clusterNotes, err
+}
+
+func (impl ClusterNoteRepositoryImpl) Update(model *ClusterNote) error {
+	return impl.dbConnection.Update(model)
+}
+
+func (impl ClusterNoteRepositoryImpl) Delete(model *ClusterNote) error {
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply pkg/cluster/repository/ClusterNoteRepository.go]: removed delete method 👍

File: pkg/cluster/repository/ClusterNoteHistoryRepository.go (hunk 2 above)
Comment (Ash-exp) [reply pkg/cluster/repository/ClusterNoteHistoryRepository.go:55]: Implemented  👍

File: pkg/cluster/ClusterNoteService.go (hunk 3 above)
Comment (Ash-exp) [reply pkg/cluster/ClusterNoteService.go]: removed delete method 👍

File: pkg/cluster/repository/ClusterNoteRepository.go (hunk 1 above)
Comment (Ash-exp) [reply pkg/cluster/repository/ClusterNoteRepository.go]: Updated 👍

File: pkg/cluster/ClusterNoteService.go (hunk 4 above)
Comment (Ash-exp) [reply pkg/cluster/ClusterNoteService.go:182]: created by and created on is required for saving the record to the audit history

File: pkg/cluster/ClusterNoteService.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	clusterAudit.UpdatedOn = time.Now()
+	err = impl.clusterNoteHistoryRepository.SaveHistory(clusterAudit)
+	if err != nil {
+		return nil, err
```
Comment (iamayushm) [inline pkg/cluster/ClusterNoteService.go:155]: Don't return err, only log it if there is error in saving history

File: api/cluster/ClusterRestHandler.go [hunk 7]
```diff
@@ -200,6 +203,58 @@ func (impl ClusterRestHandlerImpl) FindById(w http.ResponseWriter, r *http.Reque
...
+	}
+	// RBAC enforcer applying
+	token := r.Header.Get("token")
+	if ok := impl.enforcer.Enforce(token, casbin.ResourceCluster, casbin.ActionGet, strings.ToLower(clusterBean.ClusterName)); !ok {
+		common.WriteJsonResp(w, errors.New("unauthorized"), nil, http.StatusForbidden)
+		return
+	}
+	//RBAC enforcer Ends
+	bean, err := impl.clusterNoteService.FindByClusterId(i)
+	if err != nil {
```
Comment (iamayushm) [inline api/cluster/ClusterRestHandler.go:235]: use join instead of two queries

**Review by iamayushm — APPROVED** [review]

File: api/cluster/ClusterRestHandler.go (hunk 7 above)
Comment (Ash-exp) [reply api/cluster/ClusterRestHandler.go:238]: implemented 👍

File: scripts/sql/133_alter_cluster_details.up.sql [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    "created_on"                timestamptz  NOT NULL,
+    "created_by"                int4         NOT NULL,
+    "updated_on"                timestamptz,
+    "updated_by"                int4,
+    PRIMARY KEY ("id")
+);
+
+-- add foreign key
+ALTER TABLE "public"."cluster_note"
+    ADD FOREIGN KEY ("cluster_id") REFERENCES "public"."cluster" ("id");
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply scripts/sql/133_alter_cluster_details.up.sql]: implemented 👍

File: pkg/cluster/ClusterNoteService.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+	"go.uber.org/zap"
+)
+
+type ClusterNoteBean struct {
+	Id          int       `json:"id" validate:"number"`
+	ClusterId   int       `json:"cluster_id" validate:"required"`
+	Description string    `json:"description" validate:"required"`
+	UpdatedBy   int       `json:"updated_by"`
+	UpdatedOn   time.Time `json:"updated_on"`
+}
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply pkg/cluster/ClusterNoteService.go:37]: yeah this is required for unmarshalling of ClusterNoteRepository

File: pkg/cluster/ClusterNoteService.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+		NoteId:      bean.Id,
+		Description: bean.Description,
+	}
+	clusterAudit.CreatedBy = userId
+	clusterAudit.UpdatedBy = userId
+	clusterAudit.CreatedOn = time.Now()
+	clusterAudit.UpdatedOn = time.Now()
+	err = impl.clusterNoteHistoryRepository.SaveHistory(clusterAudit)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply pkg/cluster/ClusterNoteService.go]: implemented 👍

File: pkg/cluster/ClusterNoteService.go (hunk 6 above)
Comment (Ash-exp) [reply pkg/cluster/ClusterNoteService.go]: implemented 👍

**Review by iamayushm — APPROVED** [review]

**Pull request merged**